helix-term/src/commands.rs

```diff
@@ -2404,18 +2407,41 @@ fn global_search(cx: &mut Context) {
     struct GlobalSearchConfig {
         smart_case: bool,
         file_picker_config: helix_view::editor::FilePickerConfig,
+        directory_style: Style,
+        number_style: Style,
+        colon_style: Style,
     }
 
     let config = cx.editor.config();
     let config = GlobalSearchConfig {
         smart_case: config.search.smart_case,
         file_picker_config: config.file_picker.clone(),
+        directory_style: cx.editor.theme.get("ui.text.directory"),
+        number_style: cx.editor.theme.get("constant.numeric"),
+        colon_style: cx.editor.theme.get("punctuation"),
     };
 
     let columns = [
-        PickerColumn::new("path", |item: &FileResult, _| {
+        PickerColumn::new("path", |item: &FileResult, config: &GlobalSearchConfig| {
             let path = helix_stdx::path::get_relative_path(&item.path);
-            format!("{}:{}", path.to_string_lossy(), item.line_num + 1).into()
+
+            let directories = path
+                .parent()
+                .filter(|p| !p.as_os_str().is_empty())
+                .map(|p| p.to_string_lossy().to_string() + "/")
+                .unwrap_or_default();
```

Comment:
```suggestion
            let directories = path
                .parent()
                .filter(|p| !p.as_os_str().is_empty())
                .map(|p| format!("{}{}", p.display(), std::path::MAIN_SEPARATOR))
                .unwrap_or_default();
```

Also, I wonder, do we need the owned `to_string()` allocations ? `Span` contains a `Cow<str>` so can we get away with less allocations ?

Reply:
The following will handle splitting the directory and file, do no extra allocations (beyond `to_string_lossy`), and will handle the condition when the directory portion is empty or root.

I haven't tested on Windows. The `p.as_os_str() != std::path::MAIN_SEPARATOR_STR` should probably be an ends-with check instead.

I also noticed that `Path` has a `display` method. Any idea why the original code used `to_string_lossy` instead of `display`?

```rust
    let p = std::path::Path::new("/path/f.txt");

    let parent = p.parent()
        .filter(|p| !p.as_os_str().is_empty() && p.as_os_str() != std::path::MAIN_SEPARATOR_STR);
    let directory = parent
        .map(std::path::Path::to_string_lossy)
        .unwrap_or_default();
    let s = parent
        .and(Some(std::path::MAIN_SEPARATOR_STR))
        .unwrap_or_default();
    let file = p
        .file_name()
        .map(std::ffi::OsStr::to_string_lossy)
        .unwrap_or_default();
    let line = 45;

    println!("{directory}{s}{file}:{line}");
```

Reply:
> Also, I wonder, do we need the owned `to_string()` allocations ? `Span` contains a `Cow<str>` so can we get away with less allocations ?

That works! Also good that it considers Windows now as well.

Reply:
> The following will handle splitting the directory and file, do no extra allocations (beyond `to_string_lossy`), and will handle the condition when the directory portion is empty or root.
> 
> I haven't tested on Windows. The `p.as_os_str() != std::path::MAIN_SEPARATOR_STR` should probably be an ends-with check instead.
> 
> I also noticed that `Path` has a `display` method. Any idea why the original code used `to_string_lossy` instead of `display`?
> 
> ```rust
>     let p = std::path::Path::new("/path/f.txt");
> 
>     let parent = p.parent()
>         .filter(|p| !p.as_os_str().is_empty() && p.as_os_str() != std::path::MAIN_SEPARATOR_STR);
>     let directory = parent
>         .map(std::path::Path::to_string_lossy)
>         .unwrap_or_default();
>     let s = parent
>         .and(Some(std::path::MAIN_SEPARATOR_STR))
>         .unwrap_or_default();
>     let file = p
>         .file_name()
>         .map(std::ffi::OsStr::to_string_lossy)
>         .unwrap_or_default();
>     let line = 45;
> 
>     println!("{directory}{s}{file}:{line}");
> ```

Looks like we must split the `path` into 2 owned variables (corresponding to the filename itself, as well as the parents of the file). We can't return non-owned `Cow`s from the closure from my testing